(db) basic library structure with unit tests

Xertes0 opened the pull request.

Xertes0 committed:
[db] Basic library structure with unit test

Introduce cmake targets for database library and unit tests.
The library only contains an example.hh header with return_seven()
function to demonstrate unit tests usage.

Xertes0 committed:
[db] Add context class for sqlite

Adds context class which holds sqlite3 handle and allows for raw
queries. Includes unit tests.

Removes the example header with return_seven() function and related unit
test.

patkaczm commented on src/database/src/context.cc:
```diff
@@ -0,0 +1,43 @@
+{
+
+context::context(std::filesystem::path const& path) :
+    handle_{}
```
lets explicitly initialize `handle_` with `nullptr`

patkaczm commented on src/database/src/context.cc:
```diff
@@ -0,0 +1,43 @@
+    if(res != SQLITE_OK) {
+        std::fprintf(stderr, "%s\n", ::sqlite3_errmsg(handle_));
+        sqlite3_close_v2(handle_);
+        throw std::runtime_error{"Failed to open sqlite3 database."};
```
lets create a new exception derived from `std::runtime_error` like `db::CannotOpenDatabase` or sth similar

patkaczm commented on src/database/src/context.cc:
```diff
@@ -0,0 +1,43 @@
+    if(res != SQLITE_OK) {
+        std::fprintf(stderr, "%s\n", err_msg);
+        sqlite3_free(err_msg);
+        sqlite3_close_v2(handle_);
```
I think this is not needed since handle_ is closed anyway in the destructor

patkaczm commented on src/database/src/context.cc:
```diff
@@ -0,0 +1,43 @@
+        sqlite3_free(err_msg);
+        sqlite3_close_v2(handle_);
+        // Failed query is considered a bug and should not happen;
+        throw std::runtime_error{"Raw query failed."};
```
lets create new type of exception like `db::QueryFailed` or sth similar

patkaczm commented on src/database/src/context.cc:
```diff
@@ -0,0 +1,43 @@
+    char* err_msg = nullptr;
+    auto res = ::sqlite3_exec(handle_, query, callback, user_data, &err_msg);
+    if(res != SQLITE_OK) {
+        std::fprintf(stderr, "%s\n", err_msg);
```
how about streams? Maybe lets use `std::cerr`

patkaczm commented on src/database/include/database/context.hh:
```diff
@@ -0,0 +1,63 @@
+    void
+    raw_query(
+        char const* query,
+        int(*callback)(void*,int,char**,char**),
```
let's create using for this function pointer, sth like: 
`using Callback = int(*)(void*,int,char**,char**);`
and use it instead of pure type

patkaczm commented on src/database/include/database/context.hh:
```diff
@@ -0,0 +1,63 @@
+    ~context();
+
+    context(context const&) = delete;
+    context(context&&) = default;
```
I think that the default move constructor and assignment operator do not exactly do what you intended.
It can be said that the `defaults` performs `std::move` on each class member during `context(context&&)` or `context& operator=(context&&)` but `pointer` has no `default` `move` operation implemented since (I think) the value of the `handle_` is going to be simply copied thus we perform `shallow copy` in a place where `deep copy` should be performed which is an error in my opinion.

patkaczm commented on test/database/src/context.cc:
```diff
@@ -0,0 +1,31 @@
+
+#include <cstring>
+
+#include <database/context.hh>
```
let's use "" instead of <> since this is out 'local' implementation, not external one.

patkaczm commented on test/database/src/context.cc:
```diff
@@ -0,0 +1,31 @@
+
+TEST(Database, Context)
+{
+    db::context ctx{DATABASE_DIR "/context.sqlite"};
```
whare is `DATABASE_DIR` defined?

patkaczm commented on src/database/include/database/context.hh:
```diff
@@ -0,0 +1,63 @@
+
+#include <filesystem>
+#include <functional>
+#include <string_view>
```
remove as it is unused

patkaczm commented on src/database/include/database/context.hh:
```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <filesystem>
+#include <functional>
```
this is also unused

patkaczm commented on src/database/include/database/context.hh:
```diff
@@ -0,0 +1,63 @@
+#pragma once
```
In general, there is no need to have separated directories for `include` and `source` but if you want to keep files in this way there is also no problem.

Xertes0 committed:
[db] Cleanup context class and unit tests

 - Fix move constructor
 - Explicitly initialize handle_ with nullptr in the constructor
 - Throw exceptions derived from std::runtime_error
 - Remove unnecessary sqlite3_close_v2() calls
 - Use streams instead of std::fprintf()
 - Type alias for raw_query() callback
 - Remove unused headers

 - Use "" for including db headers from unit tests
 - Comment on DATABASE_DIR define
